kyaml/yaml/rnode.go

```diff
@@ -248,6 +388,511 @@ func (rn *RNode) IsNilOrEmpty() bool {
 		IsYNodeZero(rn.YNode())
 }
 
+// AsOrDie converts a RNode to the desired typed object. ptr must
+// be a pointer to a typed object. It will panic if it encounters an error.
+func (rn *RNode) AsOrDie(ptr interface{}) {
+	if err := rn.As(ptr); err != nil {
+		panic(err)
+	}
+}
+
+// As converts a RNode to the desired typed object. ptr must be
+// a pointer to a typed object.
+func (rn *RNode) As(ptr interface{}) error {
+	if rn == nil {
+		return fmt.Errorf("the object doesn't exist")
+	}
+	if ptr == nil || reflect.ValueOf(ptr).Kind() != reflect.Ptr {
+		return fmt.Errorf("ptr must be a pointer to an object")
+	}
+
+	// The built-in types only have json tags. We can't simply do mv.Node().Decode(ptr),
+	// since it use the lowercased field name by default if no yaml tag is specified.
+	// This affects both k8s built-in types (e.g. appsv1.Deployment) and any types
+	// that depends on built-in types (e.g. metav1.ObjectMeta, corev1.PodTemplate).
+	// To work around it, we rely on the json tags. We first convert mv to json
+	// and then unmarshal it to ptr.
```

Comment:
To work around this problem, we can change the yaml decoding logic. We can fall back to use json tag if yaml tag is not available. It makes sense in practice.
@KnVerey WDYT?

Reply:
Quote from https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/739#issue-1102600327

> Use sigs.k8s.io/yaml for yaml formatting, to avoid a few bugs in sigs.k8s.io/kustomize/kyaml/yaml

`sigs.k8s.io/yaml` relies on json tag. And users consider it as a bug in sigs.k8s.io/kustomize/kyaml/yaml.